File: games/mindustry/js/combat.js
/** 波次、敌人、炮塔射击与子弹 */

import { WAVE, MAP_W, MAP_H } from './config.js';
import { findPath, getTile } from './world.js';
import { isSolidBuilding, damageBuilding, buildingAt } from './buildings.js';

let enemyId = 1;
let bulletId = 1;

export function createCombatState() {
  return {
    enemies: [],
    bullets: [],
    wave: 0,
    waveTimer: WAVE.firstDelay,
    betweenWaves: true,
    spawnQueue: 0,
    spawnAcc: 0,
    pathCacheTimer: 0,
  };
}

function isBlocked(world, x, y) {
  const t = getTile(world, x, y);
  if (!t) return true;
  return isSolidBuilding(t.building);
}

function spawnPoint() {
  // 从地图边缘随机刷怪
  const edge = Math.floor(Math.random() * 4);
  if (edge === 0) return { x: Math.random() * MAP_W, y: 0.5 };
  if (edge === 1) return { x: Math.random() * MAP_W, y: MAP_H - 0.5 };
  if (edge === 2) return { x: 0.5, y: Math.random() * MAP_H };
  return { x: MAP_W - 0.5, y: Math.random() * MAP_H };
}

export function tickWaves(combat, core, dt) {
  if (!core || core.hp <= 0) return;

  if (combat.spawnQueue > 0) {
    combat.spawnAcc += dt;
    if (combat.spawnAcc >= 0.45) {
      combat.spawnAcc = 0;
      combat.spawnQueue--;
      spawnEnemy(combat, core);
    }
  }

  if (!combat.betweenWaves) {
    if (combat.spawnQueue <= 0 && combat.enemies.length === 0) {
      combat.betweenWaves = true;
      combat.waveTimer = WAVE.interval;
    }
    return;
  }

  combat.waveTimer -= dt;
  if (combat.waveTimer <= 0) {
    combat.wave++;
    combat.betweenWaves = false;
    combat.spawnQueue = WAVE.baseCount + (combat.wave - 1) * WAVE.countGrowth;
    combat.spawnAcc = 0;
  }
}

function spawnEnemy(combat, core) {
  const p = spawnPoint();
  const w = combat.wave;
  const hp = WAVE.hpBase + (w - 1) * WAVE.hpGrowth;
  const speed = WAVE.speedBase + (w - 1) * WAVE.speedGrowth;
  combat.enemies.push({
    id: enemyId++,
    x: p.x,
    y: p.y,
    hp,
    maxHp: hp,
    speed: Math.min(speed, 2.4),
    damage: WAVE.damage + Math.floor(w / 3),
    attackAcc: 0,
    path: null,
    pathIdx: 0,
    retarget: 0,
    kind: w >= 5 && Math.random() < 0.25 ? 'tank' : 'crawler',
  });

  // tank 变体更肉
  const e = combat.enemies[combat.enemies.length - 1];
  if (e.kind === 'tank') {
    e.hp *= 2.2;
    e.maxHp = e.hp;
    e.speed *= 0.65;
    e.damage *= 1.4;
  }
}

export function tickEnemies(combat, state, world, inventory, core, dt) {
  if (!core) return { coreDead: false };

  combat.pathCacheTimer -= dt;
  let coreDead = false;

  for (let i = combat.enemies.length - 1; i >= 0; i--) {
    const e = combat.enemies[i];
    if (e.hp <= 0) {
      combat.enemies.splice(i, 1);
      // 击杀奖励
      core.storage.copper = (core.storage.copper || 0) + 2;
      if (Math.random() < 0.35) core.storage.lead = (core.storage.lead || 0) + 1;
      continue;
    }

    e.retarget -= dt;
    const cx = core.x + core.def.size / 2;
    const cy = core.y + core.def.size / 2;

    // 近战攻击核心或挡路建筑
    const distCore = Math.hypot(e.x - cx, e.y - cy);
    if (distCore < core.def.size * 0.65 + 0.35) {
      e.attackAcc += dt;
      if (e.attackAcc >= WAVE.attackInterval) {
        e.attackAcc = 0;
        coreDead = damageBuilding(state, world, inventory, core, e.damage) || coreDead;
      }
      continue;
    }

    // 攻击挡路固体建筑
    const fx = Math.floor(e.x);
    const fy = Math.floor(e.y);
    const ahead = sampleBlockAhead(world, e, cx, cy);
    if (ahead) {
      e.attackAcc += dt;
      if (e.attackAcc >= WAVE.attackInterval) {
        e.attackAcc = 0;
        damageBuilding(state, world, inventory, ahead, e.damage);
        e.path = null;
      }
      continue;
    }

    if (!e.path || e.retarget <= 0) {
      e.path = findPath(world, e.x, e.y, cx, cy, (x, y) => isBlocked(world, x, y));
      e.pathIdx = 0;
      e.retarget = 1.2 + Math.random() * 0.6;
      // 无路则直线冲向核心（拆墙）
      if (!e.path) {
        const ang = Math.atan2(cy - e.y, cx - e.x);
        e.x += Math.cos(ang) * e.speed * dt;
        e.y += Math.sin(ang) * e.speed * dt;
        continue;
      }
    }

    // 沿路径移动
    while (e.path && e.pathIdx < e.path.length - 1) {
      const node = e.path[e.pathIdx + 1];
      const tx = node.x + 0.5;
      const ty = node.y + 0.5;
      const dx = tx - e.x;
      const dy = ty - e.y;
      const dist = Math.hypot(dx, dy);
      if (dist < 0.08) {
        e.pathIdx++;
        continue;
      }
      const step = e.speed * dt;
      if (step >= dist) {
        e.x = tx;
        e.y = ty;
        e.pathIdx++;
      } else {
        e.x += (dx / dist) * step;
        e.y += (dy / dist) * step;
      }
      break;
    }

    // 轻微推挤，避免重叠
    for (let j = 0; j < i; j++) {
      const o = combat.enemies[j];
      const ddx = e.x - o.x;
      const ddy = e.y - o.y;
      const d = Math.hypot(ddx, ddy);
      if (d > 0 && d < 0.55) {
        const push = (0.55 - d) * 0.5;
        e.x += (ddx / d) * push;
        e.y += (ddy / d) * push;
      }
    }
  }

  return { coreDead };
}

function sampleBlockAhead(world, e, cx, cy) {
  const ang = Math.atan2(cy - e.y, cx - e.x);
  const px = Math.floor(e.x + Math.cos(ang) * 0.55);
  const py = Math.floor(e.y + Math.sin(ang) * 0.55);
  const b = buildingAt(world, px, py);
  if (b && isSolidBuilding(b) && b.type !== 'core') return b;
  // 脚下固体也拆
  const under = buildingAt(world, Math.floor(e.x), Math.floor(e.y));
  if (under && isSolidBuilding(under) && under.type !== 'core') return under;
  return null;
}

export function tickTurrets(state, combat, dt) {
  for (const b of state.list) {
    if (b.type !== 'duo' && b.type !== 'scatter') continue;
    b.cool = Math.max(0, b.cool - dt);

    const bx = b.x + 0.5;
    const by = b.y + 0.5;
    const range = b.def.range;
    let target = null;
    let best = range;

    for (const e of combat.enemies) {
      const d = Math.hypot(e.x - bx, e.y - by);
      if (d <= best) {
        best = d;
        target = e;
      }
    }
    if (!target) continue;

    b.aim = Math.atan2(target.y - by, target.x - bx);
    if (b.cool > 0) continue;
    if ((b.ammoBuffer || 0) < b.def.ammoPerShot) continue;

    b.ammoBuffer -= b.def.ammoPerShot;
    b.cool = 1 / b.def.fireRate;

    const pellets = b.def.pellets || 1;
    for (let i = 0; i < pellets; i++) {
      const spread = pellets > 1 ? (i - (pellets - 1) / 2) * 0.18 : 0;
      const ang = b.aim + spread;
      combat.bullets.push({
        id: bulletId++,
        x: bx + Math.cos(ang) * 0.35,
        y: by + Math.sin(ang) * 0.35,
        vx: Math.cos(ang) * 14,
        vy: Math.sin(ang) * 14,
        dmg: b.def.damage,
        life: 0.7,
        from: 'turret',
      });
    }
  }
}

export function tickBullets(combat, dt) {
  for (let i = combat.bullets.length - 1; i >= 0; i--) {
    const b = combat.bullets[i];
    b.x += b.vx * dt;
    b.y += b.vy * dt;
    b.life -= dt;
    if (b.life <= 0) {
      combat.bullets.splice(i, 1);
      continue;
    }

    let hit = false;
    for (const e of combat.enemies) {
      if (Math.hypot(e.x - b.x, e.y - b.y) < 0.38) {
        e.hp -= b.dmg;
        hit = true;
        break;
      }
    }
    if (hit) combat.bullets.splice(i, 1);
  }
}

export function playerShoot(combat, player, aimAng) {
  combat.bullets.push({
    id: bulletId++,
    x: player.x + Math.cos(aimAng) * 0.4,
    y: player.y + Math.sin(aimAng) * 0.4,
    vx: Math.cos(aimAng) * 12,
    vy: Math.sin(aimAng) * 12,
    dmg: player.damage,
    life: 0.55,
    from: 'player',
  });
}

File: steel-core/src/systems/combat.rs
use bevy::prelude::*;
use rand::Rng;

use crate::components::*;
use crate::config::{colors, tile_to_world, world_to_tile, BuildingKind, WaveStats, TILE, MAP_H, MAP_W};
use crate::map::{GameMap, GameOver, Inventory, WaveState};

pub fn tick_waves(
    time: Res<Time>,
    over: Res<GameOver>,
    mut wave: ResMut<WaveState>,
    mut commands: Commands,
    core_q: Query<&Building, With<CoreTag>>,
    enemies: Query<Entity, With<Enemy>>,
) {
    if over.done || core_q.is_empty() {
        return;
    }
    let dt = time.delta_secs();

    if wave.spawn_queue > 0 {
        wave.spawn_acc += dt;
        if wave.spawn_acc >= 0.45 {
            wave.spawn_acc = 0.0;
            wave.spawn_queue -= 1;
            spawn_enemy(&mut commands, wave.wave);
        }
    }

    if !wave.between {
        if wave.spawn_queue <= 0 && enemies.iter().count() == 0 {
            wave.between = true;
            wave.timer = WaveStats::INTERVAL;
        }
        return;
    }

    wave.timer -= dt;
    if wave.timer <= 0.0 {
        wave.wave += 1;
        wave.between = false;
        wave.spawn_queue = WaveStats::BASE_COUNT + (wave.wave as i32 - 1) * WaveStats::COUNT_GROWTH;
        wave.spawn_acc = 0.0;
    }
}

fn spawn_enemy(commands: &mut Commands, wave: u32) {
    let mut rng = rand::thread_rng();
    let edge = rng.gen_range(0..4);
    let (tx, ty) = match edge {
        0 => (rng.gen_range(0..MAP_W), MAP_H - 1),
        1 => (rng.gen_range(0..MAP_W), 0),
        2 => (0, rng.gen_range(0..MAP_H)),
        _ => (MAP_W - 1, rng.gen_range(0..MAP_H)),
    };
    let is_tank = wave >= 5 && rng.gen_bool(0.25);
    let mut hp = WaveStats::HP_BASE + (wave.saturating_sub(1) as f32) * WaveStats::HP_GROWTH;
    let mut speed = WaveStats::SPEED_BASE + (wave.saturating_sub(1) as f32) * WaveStats::SPEED_GROWTH;
    let mut damage = WaveStats::DAMAGE + (wave / 3) as f32;
    if is_tank {
        hp *= 2.2;
        speed *= 0.65;
        damage *= 1.4;
    }
    let pos = tile_to_world(tx, ty, 1);
    let color = if is_tank {
        colors::ENEMY_TANK
    } else {
        colors::ENEMY
    };
    let size = if is_tank {
        Vec2::new(22.0, 18.0)
    } else {
        Vec2::new(16.0, 14.0)
    };

    let e = commands
        .spawn((
            Sprite {
                color,
                custom_size: Some(size),
                ..default()
            },
            Transform::from_translation(pos + Vec3::new(0.0, 0.0, 4.5)),
            Enemy {
                hp,
                max_hp: hp,
                speed,
                damage,
                attack_acc: 0.0,
                path: Vec::new(),
                path_idx: 0,
                retarget: 0.0,
                is_tank,
            },
        ))
        .id();
    crate::systems::setup::spawn_hp_bar(commands, e, size.x + 8.0);
}

pub fn tick_enemies(
    time: Res<Time>,
    over: Res<GameOver>,
    mut map: ResMut<GameMap>,
    mut inv: ResMut<Inventory>,
    mut commands: Commands,
    mut enemies: Query<(Entity, &mut Transform, &mut Enemy)>,
    mut buildings: Query<&mut Building>,
    cores: Query<Entity, With<CoreTag>>,
) {
    if over.done {
        return;
    }
    let dt = time.delta_secs();
    let Ok(core_e) = cores.get_single() else {
        return;
    };
    let Ok(core_b) = buildings.get(core_e) else {
        return;
    };
    let goals: Vec<(i32, i32)> = {
        let s = core_b.kind.size();
        let mut g = Vec::new();
        for y in core_b.ty..core_b.ty + s {
            for x in core_b.tx..core_b.tx + s {
                g.push((x, y));
            }
        }
        g
    };
    let core_pos = tile_to_world(core_b.tx, core_b.ty, core_b.kind.size()).truncate();
    let core_size = core_b.kind.size();

    let mut killed = Vec::new();
    let mut core_hits = 0.0_f32;
    let mut building_damage: Vec<(Entity, f32)> = Vec::new();

    for (e, mut tf, mut enemy) in enemies.iter_mut() {
        if enemy.hp <= 0.0 {
            killed.push(e);
            continue;
        }

        let pos = tf.translation.truncate();
        if pos.distance(core_pos) < core_size as f32 * TILE * 0.55 {
            enemy.attack_acc += dt;
            if enemy.attack_acc >= WaveStats::ATTACK_INTERVAL {
                enemy.attack_acc = 0.0;
                core_hits += enemy.damage;
            }
            continue;
        }

        // 拆挡路建筑
        let (ex, ey) = world_to_tile(pos);
        let ang = (core_pos - pos).normalize_or_zero();
        let ahead = world_to_tile(pos + ang * (TILE * 0.6));
        let block = [ahead, (ex, ey)]
            .into_iter()
            .find_map(|(x, y)| {
                map.get(x, y).and_then(|t| {
                    t.building.filter(|_| t.solid).filter(|&b| b != core_e)
                })
            });
        if let Some(be) = block {
            enemy.attack_acc += dt;
            if enemy.attack_acc >= WaveStats::ATTACK_INTERVAL {
                enemy.attack_acc = 0.0;
                building_damage.push((be, enemy.damage));
                enemy.path.clear();
            }
            continue;
        }

        enemy.retarget -= dt;
        if enemy.path.is_empty() || enemy.retarget <= 0.0 {
            let path = map.find_path(ex, ey, &goals).unwrap_or_default();
            enemy.path = path;
            enemy.path_idx = 0;
            enemy.retarget = 1.2;
            if enemy.path.is_empty() {
                tf.translation.x += ang.x * enemy.speed * dt;
                tf.translation.y += ang.y * enemy.speed * dt;
                continue;
            }
        }

        if enemy.path_idx + 1 < enemy.path.len() {
            let (nx, ny) = enemy.path[enemy.path_idx + 1];
            let target = tile_to_world(nx, ny, 1).truncate();
            let delta = target - pos;
            let dist = delta.length();
            if dist < 4.0 {
                enemy.path_idx += 1;
            } else {
                let step = enemy.speed * dt;
                let dir = delta / dist;
                tf.translation.x += dir.x * step;
                tf.translation.y += dir.y * step;
            }
        }
    }

    for e in killed {
        inv.copper += 2;
        if rand::random::<f32>() < 0.35 {
            inv.lead += 1;
        }
        commands.entity(e).despawn_recursive();
    }

    if core_hits > 0.0 {
        if let Ok(mut b) = buildings.get_mut(core_e) {
            b.hp -= core_hits;
        }
    }
    for (be, dmg) in building_damage {
        if let Ok(mut b) = buildings.get_mut(be) {
            b.hp -= dmg;
            if b.hp <= 0.0 && b.kind != BuildingKind::Core {
                let (tx, ty, size) = (b.tx, b.ty, b.kind.size());
                map.clear(tx, ty, size);
                commands.entity(be).despawn_recursive();
            }
        }
    }
}

pub fn tick_turrets(
    time: Res<Time>,
    over: Res<GameOver>,
    mut commands: Commands,
    mut turrets: Query<(&Transform, &mut Turret, &Building)>,
    enemies: Query<&Transform, With<Enemy>>,
) {
    if over.done {
        return;
    }
    let dt = time.delta_secs();
    for (tf, mut turret, _) in turrets.iter_mut() {
        turret.cool = (turret.cool - dt).max(0.0);
        let origin = tf.translation.truncate();
        let mut best: Option<(Vec2, f32)> = None;
        for et in enemies.iter() {
            let p = et.translation.truncate();
            let d = p.distance(origin);
            if d <= turret.range {
                if best.map(|(_, bd)| d < bd).unwrap_or(true) {
                    best = Some((p, d));
                }
            }
        }
        let Some((target, _)) = best else {
            continue;
        };
        let aim = (target - origin).normalize_or_zero();
        turret.aim = aim.y.atan2(aim.x);
        if turret.cool > 0.0 || turret.ammo < turret.ammo_cost {
            continue;
        }
        turret.ammo -= turret.ammo_cost;
        turret.cool = 1.0 / turret.fire_rate;

        for i in 0..turret.pellets {
            let spread = if turret.pellets > 1 {
                (i as f32 - (turret.pellets as f32 - 1.0) * 0.5) * 0.18
            } else {
                0.0
            };
            let a = turret.aim + spread;
            let dir = Vec2::new(a.cos(), a.sin());
            let spawn = origin + dir * 12.0;
            commands.spawn((
                Sprite {
                    color: colors::BULLET,
                    custom_size: Some(Vec2::splat(5.0)),
                    ..default()
                },
                Transform::from_translation(spawn.extend(6.0)),
                Bullet {
                    damage: turret.damage,
                    life: 0.7,
                },
                LinearVelocity(dir * 480.0),
            ));
        }
    }
}

pub fn tick_bullets(
    time: Res<Time>,
    mut commands: Commands,
    mut bullets: Query<(Entity, &mut Transform, &mut Bullet, &LinearVelocity)>,
    mut enemies: Query<(Entity, &Transform, &mut Enemy), Without<Bullet>>,
) {
    let dt = time.delta_secs();
    for (e, mut tf, mut bullet, vel) in bullets.iter_mut() {
        tf.translation.x += vel.0.x * dt;
        tf.translation.y += vel.0.y * dt;
        bullet.life -= dt;
        if bullet.life <= 0.0 {
            commands.entity(e).despawn();
            continue;
        }
        let bp = tf.translation.truncate();
        for (ee, et, mut enemy) in enemies.iter_mut() {
            if et.translation.truncate().distance(bp) < 12.0 {
                enemy.hp -= bullet.damage;
                commands.entity(e).despawn();
                if enemy.hp <= 0.0 {
                    // 击杀掉落在 tick_enemies 处理；此处仅扣血
                    let _ = ee;
                }
                break;
            }
        }
    }
}

pub fn update_hp_bars(
    mut bars: Query<(&HpBar, &mut Transform, &mut Sprite)>,
    targets: Query<(Entity, &Transform, Option<&Building>, Option<&Enemy>, Option<&Player>), Without<HpBar>>,
) {
    for (e, tf, building, enemy, player) in targets.iter() {
        let ratio = if let Some(b) = building {
            b.hp / b.max_hp
        } else if let Some(en) = enemy {
            en.hp / en.max_hp
        } else if let Some(p) = player {
            p.hp / p.max_hp
        } else {
            continue;
        };

        for (bar, mut btf, mut sprite) in bars.iter_mut() {
            if bar.owner != e {
                continue;
            }
            btf.translation.x = tf.translation.x;
            btf.translation.y = tf.translation.y + 18.0;
            btf.translation.z = 8.0;
            sprite.custom_size = Some(Vec2::new(bar.width * ratio.clamp(0.0, 1.0), 3.0));
            sprite.color = if ratio > 0.35 {
                colors::HP_GOOD
            } else {
                colors::HP_BAD
            };
        }
    }
}

pub fn check_game_over(
    mut over: ResMut<GameOver>,
    wave: Res<WaveState>,
    cores: Query<&Building, With<CoreTag>>,
    enemies: Query<Entity, With<Enemy>>,
) {
    if over.done {
        return;
    }
    if let Ok(core) = cores.get_single() {
        if core.hp <= 0.0 {
            over.done = true;
            over.won = false;
        }
    } else {
        over.done = true;
        over.won = false;
    }
    if wave.wave >= WaveStats::WIN_WAVE && wave.between && enemies.iter().count() == 0 {
        over.done = true;
        over.won = true;
    }
}
